Cleanup

Allcharles opened the pull request:
# General Cleanup

General cleanup of tests and basic components.

## Changes

- Cleaned up home component
- Cleaned up app component and tests
- Updated secondary and action menus to handle disabled links
- Implemented menu layout for Page Not Found component
- Implemented ghost users for Permissions Shield Component
- Minor cleanup to baw api services
- Merged menu components into the `/components/shared/menu` folder and created a `@menu` import path

### Test Coverage Before PR:

```
=============================== Coverage summary ===============================
TOTAL: 1809 SUCCESS
Statements   : 85.77% ( 3405/3970 )
Branches     : 81.4% ( 906/1113 )
Functions    : 75.33% ( 962/1277 )
Lines        : 85.26% ( 3152/3697 )
================================================================================
```

### Test Coverage After PR:

```
TOTAL: 1850 SUCCESS
=============================== Coverage summary ===============================
Statements   : 86.45% ( 3444/3984 )
Branches     : 83.21% ( 912/1096 )
Functions    : 75.76% ( 969/1279 )
Lines        : 85.99% ( 3191/3711 )
================================================================================
```

## Visual Changes

### Page Not Found Page

Page not found link is disabled so that users are unable to click on it, this also makes it highlighted.

![image](https://user-images.githubusercontent.com/3955116/98194702-b2ffc380-1f6b-11eb-9967-331c5e6bdce0.png)

### Regular Users in Permissions Shield Component

![image](https://user-images.githubusercontent.com/3955116/98194744-cca10b00-1f6b-11eb-8ea8-1f71ca72b700.png)

### Ghost Users in Permissions Shield Component

![image](https://user-images.githubusercontent.com/3955116/98194721-bc892b80-1f6b-11eb-9be5-69d8cdb2a4e3.png)

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Ensure project linter is not producing any warnings (`npm run lint`)
- [x] Ensure build is passing on all browsers (`npm run test:all`)
- [x] Ensure CI build is passing

Allcharles committed:
Cleanup home component

Allcharles committed:
Cleanup app component

Implemented better tests. Cleaned up the code to be more readable

Allcharles committed:
Fixed mistake in app component

App component was not setting menu layout for page components which did not define a fullscreen value

Allcharles committed:
Implemented menu layout for page not found

Page not found now will display the full menu layout. It will also display a disabled page not found link on the secondary menu. This involved modifying the internal/external/button menu items.

Allcharles committed:
Created tests for error components

Fixed broken button component unit tests

Allcharles committed:
Fixed #567 review comment

Fixed up comment left here: https://github.com/QutEcoacoustics/workbench-client/pull/567#discussion_r515775248

Allcharles committed:
Implemented ghost users

- Created unknown and deleted user accounts
- Modified permissions shield to better display ghost users
- Re-implemented user-badge component
- Merged user-badges component into permissions-shield
- Updated accounts service to create ghost users
- Merged menu components into a single folder
- Created @menu typescript import path
- Removed console warning from association decorators as it was spamming too much to be useful
- Created isGhostUser pipe to identify ghost user models

Allcharles committed:
Update services to use generators

Allcharles committed:
Linted project

atruskie commented on src/app/app.component.ts:
```diff
@@ -31,49 +40,62 @@ export class AppComponent extends WithUnsubscribe() implements OnInit {
+      count++;
+    }
+
+    if (count === 50) {
```
```suggestion
    if (count >= 50) {
```

atruskie commented on src/app/components/shared/menu/button/button.component.ts:
```diff
@@ -7,20 +7,24 @@ import { MenuAction } from "@interfaces/menusInterfaces";
+    <!--
+      No tooltips for disabled buttons because of:
+      https://github.com/ng-bootstrap/ng-bootstrap/issues/1250#issuecomment-274916839
+    -->
```
if this is needed, I believe bootstrap's solution is to use a wrapper element to which the tool tip is applied

atruskie commented on src/app/models/User.ts:
```diff
@@ -92,10 +92,24 @@ export class User extends AbstractModel implements IUser {
+      userName: "Unknown User",
+    });
+  }
+
```
these users should be constant, and frozen. Then they can be compared via reference equality.

Unless there's a specific need to return a different instance each time?

Allcharles replied:
My reasoning at the time was that it would prevent changes to the User object. However `Object.freeze()` fixes that

Allcharles replied:
As it turns out, we cannot freeze these objects. The decorator which modifies the user image urls occurs during runtime and will fail if the user object is frozen.

atruskie commented on src/app/services/baw-api/account/accounts.service.ts:
```diff
@@ -40,7 +42,21 @@ export class AccountsService extends StandardApi<User> {
+            console.warn("Returning Unknown User");
+            return of(User.unknownUser);
+          case apiReturnCodes.notFound:
+            console.warn("Returning Not Found User");
```
these probably don't need to be logged - maybe better to add comments and explain the situations for which these occur a little more.

- not found: only occurs when a user if soft-deleted
- unknown: only occurs when user is anonymous to server

Allcharles committed:
Cleanup review comments

Allcharles committed:
Cleanup menu component unit test

Extracted FontAwesome module to IconsModule. This will allow other modules to import the IconsModule instead of the full SharedModule.

Allcharles committed:
Merge branch 'master' into cleanup
